Source/WebCore/Modules/reporting/Reporting.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "Reporting.h"
+
+#include "ContextDestructionObserver.h"
+#include "HeaderFieldTokenizer.h"
+#include "ReportingObserver.h"
+#include "ScriptExecutionContext.h"
+#include <wtf/text/StringParsingBuffer.h>
+
+namespace WebCore {
+
+Ref<Reporting> Reporting::create(ScriptExecutionContext& scriptExecutionContext)
+{
+    return adoptRef(*new Reporting(scriptExecutionContext));
+}
+
+Reporting::Reporting(ScriptExecutionContext& scriptExecutionContext)
+    : ContextDestructionObserver(&scriptExecutionContext)
+{
+}
+
+Reporting::~Reporting() = default;
+
+void Reporting::registerReportingObserver(ReportingObserver& observer)
+{
+    if (!m_reportingObservers.contains(&observer))
+        m_reportingObservers.append(observer);
+}
+
+void Reporting::unregisterReportingObserver(const ReportingObserver& observer)
+{
+    m_reportingObservers.removeFirst(&observer);
+}
+
+void Reporting::removeAllObservers()
+{
+    m_reportingObservers.clear();
+}
+
+void Reporting::clearReports()
+{
+    m_queuedReports.clear();
+}
+
+void Reporting::notifyReportObservers(Ref<Report>&& report)
```

Comment:
Add a URL to https://w3c.github.io/reporting/#notify-observers?

Reply:
Done.